File: engine/src/mining/MiningSystem.cpp
#include "mining/MiningSystem.h"

namespace subspace {

MiningSystem::MiningSystem() : SystemBase("MiningSystem") {}

void MiningSystem::Update(float /*deltaTime*/) {
    // Standalone: callers drive per-entity mining/salvaging via ProcessMining/ProcessSalvaging.
}

void MiningSystem::AddAsteroid(const Asteroid& asteroid) {
    _asteroids[asteroid.id] = asteroid;
}

void MiningSystem::AddWreckage(const Wreckage& wreckage) {
    _wreckage[wreckage.id] = wreckage;
}

bool MiningSystem::StartMining(MiningComponent& miner, EntityId asteroidId,
                               const MiningPosition& minerPos) {
    auto it = _asteroids.find(asteroidId);
    if (it == _asteroids.end()) return false;

    float distance = minerPos.DistanceTo(it->second.position);
    if (distance > miner.miningRange) return false;

    miner.isMining = true;
    miner.targetAsteroidId = asteroidId;
    return true;
}

void MiningSystem::StopMining(MiningComponent& miner) {
    miner.isMining = false;
    miner.targetAsteroidId = InvalidEntityId;
}

bool MiningSystem::StartSalvaging(SalvagingComponent& salvager, EntityId wreckageId,
                                  const MiningPosition& salvagerPos) {
    auto it = _wreckage.find(wreckageId);
    if (it == _wreckage.end()) return false;

    float distance = salvagerPos.DistanceTo(it->second.position);
    if (distance > salvager.salvageRange) return false;

    salvager.isSalvaging = true;
    salvager.targetWreckageId = wreckageId;
    return true;
}

void MiningSystem::StopSalvaging(SalvagingComponent& salvager) {
    salvager.isSalvaging = false;
    salvager.targetWreckageId = InvalidEntityId;
}

float MiningSystem::ProcessMining(MiningComponent& miner, Inventory& inventory,
                                  float deltaTime) {
    if (!miner.isMining || miner.targetAsteroidId == InvalidEntityId) return 0.0f;

    auto it = _asteroids.find(miner.targetAsteroidId);
    if (it == _asteroids.end()) {
        StopMining(miner);
        return 0.0f;
    }

    Asteroid& asteroid = it->second;
    float extracted = std::min(miner.miningPower * deltaTime, asteroid.remainingResources);
    asteroid.remainingResources -= extracted;

    inventory.AddResource(asteroid.resourceType, static_cast<int>(extracted));

    // Remove asteroid if depleted
    if (asteroid.remainingResources <= 0.0f) {
        _asteroids.erase(it);
        StopMining(miner);
    }

    return extracted;
}

int MiningSystem::ProcessSalvaging(SalvagingComponent& salvager, Inventory& inventory,
                                   float deltaTime) {
    if (!salvager.isSalvaging || salvager.targetWreckageId == InvalidEntityId) return 0;

    auto it = _wreckage.find(salvager.targetWreckageId);
    if (it == _wreckage.end()) {
        StopSalvaging(salvager);
        return 0;
    }

    Wreckage& wreck = it->second;
    int salvageAmount = static_cast<int>(salvager.salvagePower * deltaTime);
    int totalSalvaged = 0;

    for (auto& [resType, remaining] : wreck.resources) {
        if (remaining <= 0) continue;
        int toSalvage = std::min(salvageAmount, remaining);
        if (inventory.AddResource(resType, toSalvage)) {
            remaining -= toSalvage;
            totalSalvaged += toSalvage;
        }
    }

    // Remove wreckage if fully salvaged
    if (wreck.IsFullySalvaged()) {
        _wreckage.erase(it);
        StopSalvaging(salvager);
    }

    return totalSalvaged;
}

const std::unordered_map<EntityId, Asteroid>& MiningSystem::GetAsteroids() const {
    return _asteroids;
}

const std::unordered_map<EntityId, Wreckage>& MiningSystem::GetWreckage() const {
    return _wreckage;
}

size_t MiningSystem::GetAsteroidCount() const { return _asteroids.size(); }
size_t MiningSystem::GetWreckageCount() const { return _wreckage.size(); }

} // namespace subspace

File: AvorionLike/Core/Mining/MiningSystem.cs
using System.Numerics;
using AvorionLike.Core.ECS;
using AvorionLike.Core.Resources;
using AvorionLike.Core.Procedural;

namespace AvorionLike.Core.Mining;

/// <summary>
/// Component for mining capabilities
/// </summary>
public class MiningComponent : IComponent
{
    public Guid EntityId { get; set; }
    public float MiningPower { get; set; } = 10f; // Resources per second
    public float MiningRange { get; set; } = 50f;
    public bool IsMining { get; set; } = false;
    public Guid? TargetAsteroidId { get; set; } = null;
}

/// <summary>
/// Component for salvaging capabilities
/// </summary>
public class SalvagingComponent : IComponent
{
    public Guid EntityId { get; set; }
    public float SalvagePower { get; set; } = 8f; // Resources per second
    public float SalvageRange { get; set; } = 50f;
    public bool IsSalvaging { get; set; } = false;
    public Guid? TargetWreckageId { get; set; } = null;
}

/// <summary>
/// Represents a mineable asteroid
/// </summary>
public class Asteroid
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Vector3 Position { get; set; }
    public float Size { get; set; }
    public ResourceType ResourceType { get; set; }
    public float RemainingResources { get; set; }
    
    public Asteroid(AsteroidData data)
    {
        Position = data.Position;
        Size = data.Size;
        // Try to parse resource type, default to Iron if invalid
        if (!Enum.TryParse<ResourceType>(data.ResourceType, out var resourceType))
        {
            resourceType = ResourceType.Iron;
        }
        ResourceType = resourceType;
        RemainingResources = Size * 10f; // Size determines total resources
    }
}

/// <summary>
/// Represents salvageable wreckage
/// </summary>
public class Wreckage
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Vector3 Position { get; set; }
    public Dictionary<ResourceType, int> Resources { get; set; } = new();
    public bool IsFullySalvaged => Resources.All(kvp => kvp.Value <= 0);
}

/// <summary>
/// System for mining and salvaging
/// </summary>
public class MiningSystem : SystemBase
{
    private readonly EntityManager _entityManager;
    private readonly Dictionary<Guid, Asteroid> _asteroids = new();
    private readonly Dictionary<Guid, Wreckage> _wreckage = new();

    public MiningSystem(EntityManager entityManager) : base("MiningSystem")
    {
        _entityManager = entityManager;
    }

    public override void Update(float deltaTime)
    {
        UpdateMining(deltaTime);
        UpdateSalvaging(deltaTime);
    }
    
    /// <summary>
    /// Add an asteroid to the sector
    /// </summary>
    public void AddAsteroid(Asteroid asteroid)
    {
        _asteroids[asteroid.Id] = asteroid;
    }
    
    /// <summary>
    /// Add wreckage to the sector
    /// </summary>
    public void AddWreckage(Wreckage wreckage)
    {
        _wreckage[wreckage.Id] = wreckage;
    }
    
    /// <summary>
    /// Start mining an asteroid
    /// </summary>
    public bool StartMining(MiningComponent miner, Guid asteroidId, Vector3 minerPosition)
    {
        if (!_asteroids.TryGetValue(asteroidId, out var asteroid))
        {
            return false;
        }
        
        // Check range
        float distance = Vector3.Distance(minerPosition, asteroid.Position);
        if (distance > miner.MiningRange)
        {
            return false;
        }
        
        miner.IsMining = true;
        miner.TargetAsteroidId = asteroidId;
        return true;
    }
    
    /// <summary>
    /// Start salvaging wreckage
    /// </summary>
    public bool StartSalvaging(SalvagingComponent salvager, Guid wreckageId, Vector3 salvagerPosition)
    {
        if (!_wreckage.TryGetValue(wreckageId, out var wreck))
        {
            return false;
        }
        
        // Check range
        float distance = Vector3.Distance(salvagerPosition, wreck.Position);
        if (distance > salvager.SalvageRange)
        {
            return false;
        }
        
        salvager.IsSalvaging = true;
        salvager.TargetWreckageId = wreckageId;
        return true;
    }
    
    /// <summary>
    /// Update all mining operations
    /// </summary>
    private void UpdateMining(float deltaTime)
    {
        var miners = _entityManager.GetAllComponents<MiningComponent>();
        
        foreach (var miner in miners)
        {
            if (!miner.IsMining || !miner.TargetAsteroidId.HasValue)
            {
                continue;
            }
            
            if (!_asteroids.TryGetValue(miner.TargetAsteroidId.Value, out var asteroid))
            {
                miner.IsMining = false;
                continue;
            }
            
            // Extract resources
            float extracted = Math.Min(miner.MiningPower * deltaTime, asteroid.RemainingResources);
            asteroid.RemainingResources -= extracted;
            
            // Add to inventory
            var inventory = _entityManager.GetComponent<InventoryComponent>(miner.EntityId);
            if (inventory != null)
            {
                inventory.Inventory.AddResource(asteroid.ResourceType, (int)extracted);
            }
            
            // Remove asteroid if depleted
            if (asteroid.RemainingResources <= 0)
            {
                _asteroids.Remove(miner.TargetAsteroidId.Value);
                miner.IsMining = false;
                miner.TargetAsteroidId = null;
            }
        }
    }
    
    /// <summary>
    /// Update all salvaging operations
    /// </summary>
    private void UpdateSalvaging(float deltaTime)
    {
        var salvagers = _entityManager.GetAllComponents<SalvagingComponent>();
        
        foreach (var salvager in salvagers)
        {
            if (!salvager.IsSalvaging || !salvager.TargetWreckageId.HasValue)
            {
                continue;
            }
            
            if (!_wreckage.TryGetValue(salvager.TargetWreckageId.Value, out var wreck))
            {
                salvager.IsSalvaging = false;
                continue;
            }
            
            // Salvage resources
            var inventory = _entityManager.GetComponent<InventoryComponent>(salvager.EntityId);
            if (inventory != null)
            {
                float salvageAmount = salvager.SalvagePower * deltaTime;
                
                foreach (var resource in wreck.Resources.Keys.ToList())
                {
                    if (wreck.Resources[resource] <= 0) continue;
                    
                    int toSalvage = Math.Min((int)salvageAmount, wreck.Resources[resource]);
                    if (inventory.Inventory.AddResource(resource, toSalvage))
                    {
                        wreck.Resources[resource] -= toSalvage;
                    }
                }
            }
            
            // Remove wreckage if fully salvaged
            if (wreck.IsFullySalvaged)
            {
                _wreckage.Remove(salvager.TargetWreckageId.Value);
                salvager.IsSalvaging = false;
                salvager.TargetWreckageId = null;
            }
        }
    }
    
    /// <summary>
    /// Get all asteroids in the sector
    /// </summary>
    public IEnumerable<Asteroid> GetAsteroids()
    {
        return _asteroids.Values;
    }
    
    /// <summary>
    /// Get all asteroids in the sector (for AI)
    /// </summary>
    public List<Asteroid> GetAllAsteroids()
    {
        return _asteroids.Values.ToList();
    }
    
    /// <summary>
    /// Get all wreckage in the sector
    /// </summary>
    public IEnumerable<Wreckage> GetWreckage()
    {
        return _wreckage.Values;
    }
}
